lib/AST/LocalizationFormat.cpp

```diff
@@ -143,30 +137,47 @@ void YAMLLocalizationProducer::forEachAvailable(
   }
 }
 
+llvm::Optional<uint32_t> getDiagnosticID(llvm::StringRef rawID) {
+  llvm::Optional<uint32_t> id;
+  auto it = llvm::find_if(diagnosticID,
+                          [&](const char *diagID) { return rawID == diagID; });
+  if (it != std::end(diagnosticID)) {
+    uint32_t index = static_cast<uint32_t>(std::distance(diagnosticID, it));
+    id.emplace(index);
+  }
+  return id;
+}
+
 template <typename T, typename Context>
 typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value, void>::type
-readYAML(llvm::yaml::IO &io, T &Seq, bool, Context &Ctx) {
+readYAML(llvm::yaml::IO &io, T &Seq, T &unknownIDs, bool, Context &Ctx) {
   unsigned count = io.beginSequence();
-  if (count)
+  if (count) {
     Seq.resize(LocalDiagID::NumDiags);
+  }
+
+  DiagnosticNode current;
   for (unsigned i = 0; i < count; ++i) {
     void *SaveInfo;
     if (io.preflightElement(i, SaveInfo)) {
-      DiagnosticNode current;
       yamlize(io, current, true, Ctx);
```

Comment:
Sorry for going back and forth here but I'm just wondering is there is a better solution then what we currently have.

Is it possible to try to do reading here directly via `preflight/postflight`  and other methods instead of calling `yamlize`? If we could do that, we could leave assumption that `id` is `uint32_t` and as a fallback add string to `unknownIDs` right? e.g.

something like - read key (`id`), read key content as enum, if that fails fallback to adding to `unknownIDs`, read next key (`msg`), read content of the (`msg`).